File: wear/src/main/java/com/asamm/locus/addon/wear/communication/WearCommService.java
package com.asamm.locus.addon.wear.communication;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.asamm.locus.addon.wear.MainApplication;
import com.asamm.locus.addon.wear.WatchDog;
import com.asamm.locus.addon.wear.common.communication.DataPath;
import com.asamm.locus.addon.wear.common.communication.LocusWearCommService;
import com.asamm.locus.addon.wear.common.communication.containers.TimeStampStorable;
import com.asamm.locus.addon.wear.common.utils.TriStateLogicEnum;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.PendingResult;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.wearable.CapabilityApi;
import com.google.android.gms.wearable.CapabilityInfo;
import com.google.android.gms.wearable.Node;
import com.google.android.gms.wearable.NodeApi;
import com.google.android.gms.wearable.Wearable;

import java.util.Iterator;
import java.util.Set;
import java.util.concurrent.TimeUnit;

/**
 * Singleton instance handling wear-device communication
 * <p>
 * Created by Milan Cejnar on 07.11.2017.
 * Asamm Software, s.r.o.
 */

public class WearCommService extends LocusWearCommService implements CapabilityApi.CapabilityListener {
	private static String TAG = WearCommService.class.getSimpleName();
	private static WearCommService mDeviceCommunicationService;

	private volatile long mLastSentDataTimestamp = 0L;

	private MainApplication mApp;

	private static final String CAPABILITY_PHONE_APP = "verify_remote_wear_for_locus_map_phone_app";

	private TriStateLogicEnum mCapableClientDetected = TriStateLogicEnum.UNKNOWN;

	private WearCommService(MainApplication c) {
		super(c);
		this.mApp = c;
	}

	public static WearCommService getInstance() {
		return mDeviceCommunicationService;
	}

	public static WearCommService initialize(final MainApplication c) {
		if (mDeviceCommunicationService == null) {
			synchronized (TAG) {
				if (mDeviceCommunicationService == null) {
					mDeviceCommunicationService = new WearCommService(c);
				}
			}
		}
		return mDeviceCommunicationService;
	}

	@Override
	protected void destroy() {
		if ((mGoogleApiClient != null) && mGoogleApiClient.isConnected()) {
			Wearable.CapabilityApi.removeCapabilityListener(
					mGoogleApiClient,
					this,
					CAPABILITY_PHONE_APP);
		}
		mApp = null;
		super.destroy();
	}

	public static void destroyInstance() {
		if (mDeviceCommunicationService != null) {
			mDeviceCommunicationService.destroy();
		}
	}

	@Override
	public void onConnected(@Nullable Bundle bundle) {
		super.onConnected(bundle);
		// Set up listeners for capability changes (install/uninstall of remote app).
		Wearable.CapabilityApi.addCapabilityListener(
				mGoogleApiClient,
				this,
				CAPABILITY_PHONE_APP);

		checkIfPhoneHasApp();

		final MainApplication app = this.mApp;
		if (app != null) {
			app.onConnected();
		}
	}

	@Override
	public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
		super.onConnectionFailed(connectionResult);
		if (mApp != null) {
			mApp.onConnectionSuspened();
		}
	}

	@Override
	public void onConnectionSuspended(int i) {
		super.onConnectionSuspended(i);
		if (mApp != null) {
			mApp.onConnectionSuspened();
		}
	}


	public void getConnectedNodes(ResultCallback<NodeApi.GetConnectedNodesResult> resultCallback) {
		Wearable.NodeApi.getConnectedNodes(mGoogleApiClient).setResultCallback(resultCallback, 1000, TimeUnit.MILLISECONDS);
	}


	@Override
	public void onCapabilityChanged(CapabilityInfo capabilityInfo) {
		Set<Node> capableNodes = capabilityInfo != null ? capabilityInfo.getNodes() : null;
		mCapableClientDetected = capableNodes != null && capableNodes.size() > 0 ?
				TriStateLogicEnum.TRUE : TriStateLogicEnum.FALSE;
		if (mCapableClientDetected == TriStateLogicEnum.TRUE) {
			Iterator<Node> nodeIt = capableNodes.iterator();
			mNodeId = null;
			while (mNodeId == null && nodeIt.hasNext()) {
				Node n = nodeIt.next();
				mNodeId = n.isNearby() ? n.getId() : null;
			}
		}
		if (mApp != null) {
			mApp.onCapableClientConnected();
		}
	}

	/**
	 * @return UNKNOWN if not known yet, TRUE if cabable client present, FALSE if capable device not found
	 */
	public TriStateLogicEnum isAppInstalledOnDevice() {
		return mCapableClientDetected;
	}

	@Override
	protected void sendDataItemWithoutConnectionCheck(DataPath path, TimeStampStorable data) {
		long currentTime = System.currentTimeMillis();
		// if keep alive command but some other command was sent recently then ignore this
		// command to save bandwidth
		if (path == DataPath.GET_KEEP_ALIVE &&
				currentTime - mLastSentDataTimestamp <= WatchDog.WD_PERIOD_TRANSMIT_KEEP_ALIVE_MS) {
			return;
		}
		// if sending other data than keep alive command, write current time to
		// postpone keep alive thread
		if (path != DataPath.GET_KEEP_ALIVE) {
			mLastSentDataTimestamp = currentTime;
		}
		super.sendDataItemWithoutConnectionCheck(path, data);
	}

	private void checkIfPhoneHasApp() {
		PendingResult<CapabilityApi.GetCapabilityResult> pendingResult =
				Wearable.CapabilityApi.getCapability(
						mGoogleApiClient,
						CAPABILITY_PHONE_APP,
						CapabilityApi.FILTER_ALL);

		pendingResult.setResultCallback((getCapabilityResult) -> {
			if (getCapabilityResult.getStatus().isSuccess()) {
				onCapabilityChanged(getCapabilityResult.getCapability());
			}
		});
	}
}

File: wear/src/main/java/com/asamm/locus/addon/wear/communication/WearCommService.kt
package com.asamm.locus.addon.wear.communication

import android.annotation.SuppressLint
import android.os.Bundle
import com.asamm.locus.addon.wear.MainApplication
import com.asamm.locus.addon.wear.WatchDog
import com.asamm.locus.addon.wear.common.communication.DataPath
import com.asamm.locus.addon.wear.common.communication.CommonCommService
import com.asamm.locus.addon.wear.common.communication.containers.TimeStampStorable
import com.asamm.locus.addon.wear.common.utils.TriStateLogicEnum
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.ResultCallback
import com.google.android.gms.wearable.CapabilityApi
import com.google.android.gms.wearable.CapabilityApi.CapabilityListener
import com.google.android.gms.wearable.CapabilityApi.GetCapabilityResult
import com.google.android.gms.wearable.CapabilityInfo
import com.google.android.gms.wearable.Node
import com.google.android.gms.wearable.NodeApi.GetConnectedNodesResult
import com.google.android.gms.wearable.Wearable
import locus.api.utils.Logger
import java.util.concurrent.TimeUnit

/**
 * Singleton instance handling wear-device communication
 *
 *
 * Created by Milan Cejnar on 07.11.2017.
 * Asamm Software, s.r.o.
 */
class WearCommService private constructor(val app: MainApplication)
    : CommonCommService(app), CapabilityListener {

    @Volatile
    private var lastSentDataTimestamp = 0L

    /**
     * @return UNKNOWN if not known yet, TRUE if cabable client present, FALSE if capable device not found
     */
    var isAppInstalledOnDevice = TriStateLogicEnum.UNKNOWN
        private set

    override fun destroy() {
        if (googleApiClient.isConnected) {
            Wearable.CapabilityApi.removeCapabilityListener(
                    googleApiClient,
                    this,
                    CAPABILITY_PHONE_APP
            )
        }
        super.destroy()
    }

    override fun onConnected(bundle: Bundle?) {
        Logger.logD("WearCommService", "onConnected($bundle)")
        super.onConnected(bundle)
        // Set up listeners for capability changes (install/uninstall of remote app).
        Wearable.CapabilityApi.addCapabilityListener(
                googleApiClient,
                this,
                CAPABILITY_PHONE_APP
        )
        checkIfPhoneHasApp()
        app.onConnected()
    }

    override fun onConnectionFailed(connectionResult: ConnectionResult) {
        Logger.logD("WearCommService", "onConnectionFailed($connectionResult)")
        super.onConnectionFailed(connectionResult)
        reconnectIfNeeded()
    }

    override fun onConnectionSuspended(i: Int) {
        Logger.logD("WearCommService", "onConnectionSuspended($i)")
        super.onConnectionSuspended(i)
        reconnectIfNeeded()
    }

    fun getConnectedNodes(resultCallback: ResultCallback<GetConnectedNodesResult>) {
        Wearable.NodeApi.getConnectedNodes(googleApiClient)
                .setResultCallback(resultCallback, 1000, TimeUnit.MILLISECONDS)
    }

    override fun onCapabilityChanged(capabilityInfo: CapabilityInfo) {
        val capableNodes = capabilityInfo.nodes
        isAppInstalledOnDevice =
                if (capableNodes.size > 0) TriStateLogicEnum.TRUE else TriStateLogicEnum.FALSE
        if (isAppInstalledOnDevice == TriStateLogicEnum.TRUE) {
            val nodeIt: Iterator<Node> = capableNodes.iterator()
            nodeId = null
            while (nodeId == null && nodeIt.hasNext()) {
                val n = nodeIt.next()
                nodeId = if (n.isNearby) n.id else null
            }
        }
        app.onConnected()
    }

    override fun sendDataItemWithoutConnectionCheck(path: DataPath, data: TimeStampStorable) {
        // if keep alive command but some other command was sent recently then ignore this
        // command to save bandwidth
        val currentTime = System.currentTimeMillis()
        if (path === DataPath.TD_KEEP_ALIVE
                && currentTime - lastSentDataTimestamp <= WatchDog.WD_PERIOD_TRANSMIT_KEEP_ALIVE_MS) {
            return
        }

        // if sending other data than keep alive command, write current time to
        // postpone keep alive thread
        if (path !== DataPath.TD_KEEP_ALIVE) {
            lastSentDataTimestamp = currentTime
        }
        super.sendDataItemWithoutConnectionCheck(path, data)
    }

    private fun checkIfPhoneHasApp() {
        val pendingResult = Wearable.CapabilityApi.getCapability(
                googleApiClient,
                CAPABILITY_PHONE_APP,
                CapabilityApi.FILTER_ALL
        )
        pendingResult.setResultCallback { getCapabilityResult: GetCapabilityResult ->
            if (getCapabilityResult.status.isSuccess) {
                onCapabilityChanged(getCapabilityResult.capability)
            }
        }
    }

    companion object {

        private const val CAPABILITY_PHONE_APP = "verify_remote_wear_for_locus_map_phone_app"

        /**
         * Instance of the wear connection service.
         */
        @JvmStatic
        @SuppressLint("StaticFieldLeak")
        lateinit var instance: WearCommService
            private set


        fun initialize(c: MainApplication) {
            if (!::instance.isInitialized) {
                instance = WearCommService(c)
            }
        }

        @JvmStatic
        fun destroyInstance() {
            instance.destroy()
        }
    }
}